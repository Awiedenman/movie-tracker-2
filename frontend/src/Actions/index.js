export const fetchInitialMovies = movies => ({
  type: 'FETCH_INITIAL_MOVIES',
  movies
});

export const userLogin = user => ({
  type: 'USER_LOGIN',
  id: user.id,
  name: user.name,
  loggedIn: user.loggedIn
});

export const userSignUp = newUserData => ({
  type: 'USER_SIGN_UP',
  id: newUserData.id
});

export const userSignOut = () => ({
  type: 'USER_SIGN_OUT'
});

export const clearUserFavorites = () => ({
  type: 'CLEAR_USER_FAVORITES'
});

export const addFavorite = movie => ({
  type: 'ADD_FAVORITE',
  favorite: movie
});

export const removeFavorite = () => ({
  type: 'REMOVE_FAVORITE'
});

export const addExistingFavorites = existingUserFavorites => {
  return {
    type: 'ADD_EXISTING_FAVORITES',
    favorites : existingUserFavorites
  };
};