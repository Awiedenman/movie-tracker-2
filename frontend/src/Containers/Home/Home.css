.home-container {
  background: #000;
  display: grid;
  grid-column-gap: 1rem;
  grid-row-gap: 1rem;
  grid-template-columns: repeat(auto-fit, minmax(27rem, 1fr));
  height: 100vh;
  overflow: hidden;
  overflow-y: scroll;
  padding: 1rem .25rem;
}