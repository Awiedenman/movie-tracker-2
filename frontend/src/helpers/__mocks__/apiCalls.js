export const postUserFavorites = jest.fn();
export const movieFetch = jest.fn();
export const fetchUserFavorites = jest.fn().mockImplementation(() =>
  Promise.resolve({
    ok: true,
    data: []
  }));
export const deleteUserFavorite = jest.fn().mockImplementation(() =>
  Promise.resolve({
    ok: true,
    json: () => Promise.resolve({status: 'success', message: 'One row was deleted'})
  }));