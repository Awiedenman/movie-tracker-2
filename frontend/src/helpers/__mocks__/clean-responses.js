export const cleanMovieResponse = jest.fn();
export const cleanFavoritesResponse = jest.fn();