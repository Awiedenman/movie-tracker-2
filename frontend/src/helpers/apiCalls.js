import { API_KEY } from '../api-key';

export const movieFetch = async () => {
  const url = `https://api.themoviedb.org/3/list/1?api_key=${API_KEY}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw Error('failed fetch request');
  }

  return await response.json();
};

export const userLoginRequest = async (email, password) => {
  const response = await fetch('http://localhost:3000/api/users',  {
    method: 'POST',
    body: JSON.stringify({ email,  password }),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (!response.ok) {
    throw Error('Email and Password do not match');
  }

  return await response.json();
};

export const userSignUpRequest = async (name, email, password) =>  {
  const response = await fetch('http://localhost:3000/api/users/new', {
    method: 'POST',
    body: JSON.stringify({ name, email, password }),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (!response.ok) {
    throw Error('Sorry, email has already been used');
  }

  return await response.json();
};

export const postUserFavorites = async (movie, userId) => {
  const { id, title, image, releaseDate, overview, average } = movie;

  const favoriteToPost = {
    movie_id: id,
    user_id: userId,
    title: title,
    poster_path: image,
    release_date: releaseDate,
    vote_average: average,
    overview: overview
  };

  const response = await fetch('http://localhost:3000/api/users/favorites/new', {
    method: 'POST',
    body: JSON.stringify(favoriteToPost),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (!response.ok) {
    throw Error('Sorry, there was problem saving you favorite. Please try again later.');
  }

  return response.json();
};

export const fetchUserFavorites = async userId => {
  const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites`);

  if (!response.ok) {
    throw Error('Sorry, we could not retrieve you favorites at this time');
  }

  return await response.json();
};

export const deleteUserfavorite = async (movieId, userId) => {
  const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites/${movieId}`, {
    method: 'POST',
    body: JSON.stringify(userId, movieId),
    headers: {
      'Content-Type': 'application/json'
    }
  });
  if (!response.ok) {
    throw Error('Sorry, we could not remove you favorite at this time.')
  }
  return await response.json();

};