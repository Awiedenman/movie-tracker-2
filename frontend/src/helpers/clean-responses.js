export const cleanInitialMovieRequest = response => {
  const { items } = response;
  return {
    movieList: items
  };
};