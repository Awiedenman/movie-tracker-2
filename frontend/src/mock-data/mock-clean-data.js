export const mockCleanedMovieList =[
  {
    average: 7.5,
    id: 284053,
    image: "/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg",
    overview: "Thor is on the other side of the universe and finds himself in a race against time to get back to Asgard to stop Ragnarok, the prophecy of destruction to his homeworld and the end of Asgardian civilization, at the hands of an all-powerful new threat, the ruthless Hela.",
    title: "Thor: Ragnarok"
  }
];