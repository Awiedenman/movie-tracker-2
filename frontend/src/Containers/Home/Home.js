import React, { Component } from 'react';
import { connect } from 'react-redux';
import { movieFetch } from '../../helpers/apiCalls';
import { cleanMovieResponse } from '../../helpers/clean-responses';
import { fetchInitialMovies, addFavorite } from '../../Actions';
import { Card } from '../../Components/Card/Card';
import PropTypes from 'prop-types';

import './Home.css';

export class Home extends Component {
  constructor() {
    super();
  }

  toggleFavorite = movie => {
    console.log('movie', movie);
    this.props.favorites(movie);

  }

  async componentDidMount () {
    const initialFetch = await movieFetch();
    const movies = cleanMovieResponse(initialFetch);
    this.props.initialFetchData(movies);
  }

  render() {
    const displayMoviesCards = this.props.movies.map(movie => (
      <Card
        {...movie}
        userId={this.props.userId}
        key={movie.id}
        toggleFavorite={this.toggleFavorite}
      />
    ));

    return (
      <div className="home-container">
        {displayMoviesCards}
      </div>
    );
  }
}

Home.propTypes = {
  initialFetchData: PropTypes.func,
  movies: PropTypes.arrayOf(PropTypes.object),
  userId: PropTypes.number,
  favorites: PropTypes.func
};

export const mapStateToProps = state => ({
  movies: state.initialMovies,
  userId: state.userInfo.id
});

export const mapDispatchToProps = dispatch => ({
  initialFetchData: movies => dispatch(fetchInitialMovies(movies)),
  favorites: movie => dispatch(addFavorite(movie))
});

export default connect(mapStateToProps, mapDispatchToProps)(Home);
