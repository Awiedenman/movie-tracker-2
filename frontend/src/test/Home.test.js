import React from 'react';
import { shallow } from 'enzyme';
import { Home, mapDispatchToProps, mapStateToProps } from '../Containers/Home/Home';
import { fetchInitialMovies, addFavorite, addExistingFavorites, removeFavorite } from '../Actions';
import { mockInitialMovieResponse } from '../mock-data/mock-responses';
import { mockCleanedMovieList } from '../mock-data/mock-clean-data';
import { postUserFavorites, fetchUserFavorites, deleteUserFavorite } from '../helpers/apiCalls';
import { cleanMovieResponse, cleanFavoritesResponse } from '../helpers/clean-responses';

jest.mock('../helpers/apiCalls');
jest.mock('../helpers/clean-responses');

describe('Home', () => {
  let wrapper;
  const mockFetchData = jest.fn();
  const mockAddFav = jest.fn();
  const mockFavFunc = jest.fn();
  const mockRemoveFav = jest.fn();
  const mockFavorites = [{ title: 'Thor', id: 2 }, { title: 'Nerds', id: 3 }];

  beforeEach(() => wrapper = shallow(
    <Home
      initialFetchData={mockFetchData}
      movies={mockCleanedMovieList}
      userId={1}
      addFavorite={mockAddFav}
      userFavorites={mockFavorites}
      getUserFavorites={mockFavFunc}
      removeFavorite={mockRemoveFav}
    />
  ));

  describe('Home', () => {
    test('should fetch initial movies when componentDidMount', async () => {
      await wrapper.instance().componentDidMount();
      cleanMovieResponse();
      fetchUserFavorites();
      cleanFavoritesResponse();
      expect(mockFetchData).toHaveBeenCalled();
    });

    test('renders without crashing', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('toggleFavorites',  () => {
    test('should call addFavorite be called', async () => {
      const mockMovie = { title: 'batman', average: 7, id: 3 };
      const mockUserId = 6;

      await wrapper.instance().toggleFavorite(mockMovie, mockUserId);
      await postUserFavorites();
      expect(mockAddFav).toHaveBeenCalled();
    });

    test('should should call removeFavorite if favorite already in fav array', async () => {
      const mockMovie = { title: 'batman', average: 7, id: 3};
      const mockUserId = 6;

      await wrapper.instance().toggleFavorite(mockMovie, mockUserId);
      await wrapper.instance().toggleFavorite(mockMovie, mockUserId);
      await removeFavorite();
      expect(mockAddFav).toHaveBeenCalled();
    });


    test('should call removeFavorite if movie is favortied', async () => {
      const mockFavorites = [{ movie_id: 1 }, { movie_id: 2}];
      const mockMovie = { id: 2 };
      const mockUserId = 6;

      wrapper = shallow(
        <Home
          initialFetchData={mockFetchData}
          movies={mockCleanedMovieList}
          userId={1}
          addFavorite={mockAddFav}
          userFavorites={mockFavorites}
          getUserFavorites={mockFavFunc}
          removeFavorite={mockRemoveFav}
        />
      );

      await wrapper.instance().toggleFavorite(mockMovie, mockUserId);
      expect(mockRemoveFav).toHaveBeenCalled();
    });

  });

  describe('mapDispatchToProps', () => {
    test('should call dispatch initialFetchData is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchInitialMovies(mockInitialMovieResponse);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.initialFetchData(mockInitialMovieResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    test('should call dispatch when addFavorites is invoked', () => {
      const mockDispatch = jest.fn();
      const mockMovie = { title: 'batman', average: 7};
      const mockUserId = 6;
      const actionToDispatch = addFavorite(mockMovie, mockUserId);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.addFavorite(mockMovie, mockUserId);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    test('should call dispatch when getUserFavorites', () => {
      const mockDispatch = jest.fn();
      const mockFavorite = { title: 'batman', average: 7 };
      const actionToDispatch = addExistingFavorites(mockFavorite);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.getUserFavorites(mockFavorite);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    test('should call dispatch when removeFavorite is invoked', () => {
      const mockDispatch = jest.fn();
      const mockMovie = { title: 'batman', average: 7 };
      const mockId = 1;
      const actionToDispatch = removeFavorite(mockMovie, mockId);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.removeFavorite(mockMovie, mockId);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

  });


  describe('mapStateToProps', () => {
    test('should return a props object with a movies array', () => {
      const expected = {
        movies: mockCleanedMovieList,
        userId: 1
      };

      const mockState = {
        initialMovies: mockCleanedMovieList,
        userInfo: {
          id: 1
        }
      };

      const actual = mapStateToProps(mockState);

      expect(actual).toEqual(expected);
    });
  });
});


