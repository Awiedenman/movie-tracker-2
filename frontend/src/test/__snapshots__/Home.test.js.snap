// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Home renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home
    initialFetchData={[Function]}
    movies={
        Array [
            Object {
              "average": 7.5,
              "id": 284053,
              "image": "/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg",
              "overview": "Thor is on the other side of the universe and finds himself in a race against time to get back to Asgard to stop Ragnarok, the prophecy of destruction to his homeworld and the end of Asgardian civilization, at the hands of an all-powerful new threat, the ruthless Hela.",
              "title": "Thor: Ragnarok",
            },
          ]
    }
    userId={1}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Card
          movie={
                    Object {
                              "average": 7.5,
                              "id": 284053,
                              "image": "/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg",
                              "overview": "Thor is on the other side of the universe and finds himself in a race against time to get back to Asgard to stop Ragnarok, the prophecy of destruction to his homeworld and the end of Asgardian civilization, at the hands of an all-powerful new threat, the ruthless Hela.",
                              "title": "Thor: Ragnarok",
                            }
          }
          toggleFavorite={[Function]}
          userId={1}
/>,
      ],
      "className": "home-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "284053",
        "nodeType": "function",
        "props": Object {
          "movie": Object {
            "average": 7.5,
            "id": 284053,
            "image": "/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg",
            "overview": "Thor is on the other side of the universe and finds himself in a race against time to get back to Asgard to stop Ragnarok, the prophecy of destruction to his homeworld and the end of Asgardian civilization, at the hands of an all-powerful new threat, the ruthless Hela.",
            "title": "Thor: Ragnarok",
          },
          "toggleFavorite": [Function],
          "userId": 1,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Card
            movie={
                        Object {
                                    "average": 7.5,
                                    "id": 284053,
                                    "image": "/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg",
                                    "overview": "Thor is on the other side of the universe and finds himself in a race against time to get back to Asgard to stop Ragnarok, the prophecy of destruction to his homeworld and the end of Asgardian civilization, at the hands of an all-powerful new threat, the ruthless Hela.",
                                    "title": "Thor: Ragnarok",
                                  }
            }
            toggleFavorite={[Function]}
            userId={1}
/>,
        ],
        "className": "home-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "284053",
          "nodeType": "function",
          "props": Object {
            "movie": Object {
              "average": 7.5,
              "id": 284053,
              "image": "/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg",
              "overview": "Thor is on the other side of the universe and finds himself in a race against time to get back to Asgard to stop Ragnarok, the prophecy of destruction to his homeworld and the end of Asgardian civilization, at the hands of an all-powerful new threat, the ruthless Hela.",
              "title": "Thor: Ragnarok",
            },
            "toggleFavorite": [Function],
            "userId": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
